src/ModelContextProtocol.Core/Server/McpServerExtensions.cs

```diff
@@ -234,6 +250,192 @@ public static ValueTask<ElicitResult> ElicitAsync(
+
+#if NET
+        var schema = dict.GetOrAdd<JsonSerializerOptions>(typeof(T), (t, s) => BuildRequestSchema(t, s), serializerOptions);
+#else
+        var schema = dict.GetOrAdd(typeof(T), type => BuildRequestSchema(type, serializerOptions));
+#endif
+
+        var request = new ElicitRequestParams
+        {
+            Message = message,
+            RequestedSchema = schema,
+        };
+
+        var raw = await server.ElicitAsync(request, cancellationToken).ConfigureAwait(false);
+
+        if (!string.Equals(raw.Action, "accept", StringComparison.OrdinalIgnoreCase) || raw.Content is null)
+        {
+            return new ElicitResult<T?> { Action = raw.Action, Content = default };
+        }
+
+        var obj = new JsonObject();
+        foreach (var kvp in raw.Content)
+        {
+            obj[kvp.Key] = JsonNode.Parse(kvp.Value.GetRawText());
+        }
+
+        T? typed = JsonSerializer.Deserialize(obj, serializerOptions.GetTypeInfo<T>());
+        return new ElicitResult<T?> { Action = raw.Action, Content = typed };
+    }
+
+    /// <summary>
+    /// Builds a request schema for elicitation based on the public serializable properties of <paramref name="type"/>.
+    /// </summary>
+    /// <param name="type">The type of the schema being built.</param>
+    /// <param name="serializerOptions">The serializer options to use.</param>
+    /// <returns>The built request schema.</returns>
+    /// <exception cref="McpException"></exception>
+    private static ElicitRequestParams.RequestSchema BuildRequestSchema(Type type, JsonSerializerOptions serializerOptions)
+    {
+        var schema = new ElicitRequestParams.RequestSchema();
+        var props = schema.Properties;
+
+        JsonTypeInfo typeInfo = serializerOptions.GetTypeInfo(type);
+
+        if (typeInfo.Kind != JsonTypeInfoKind.Object)
+        {
+            throw new McpException($"Type '{type.FullName}' is not supported for elicitation requests.");
+        }
+
+        foreach (JsonPropertyInfo pi in typeInfo.Properties)
+        {
+            var def = CreatePrimitiveSchema(pi.PropertyType, serializerOptions);
+            props[pi.Name] = def;
+        }
+
+        return schema;
+    }
+
+    /// <summary>
+    /// Creates a primitive schema definition for the specified type, if supported.
+    /// </summary>
+    /// <param name="type">The type to create the schema for.</param>
+    /// <param name="serializerOptions">The serializer options to use.</param>
+    /// <returns>The created primitive schema definition.</returns>
+    /// <exception cref="McpException">Thrown when the type is not supported.</exception>
+    private static ElicitRequestParams.PrimitiveSchemaDefinition CreatePrimitiveSchema(Type type, JsonSerializerOptions serializerOptions)
+    {
+        var originalType = type;
+        var underlyingType = Nullable.GetUnderlyingType(originalType) ?? originalType;
```

Comment:
Let's go ahead and remove this, and have nullable types fail on the `TryValidateElicitationPrimitiveSchema` method.

Reply:
> Do you mean to avoid accepting any nullable properties for `T`? It seems unusual not to accept something like:

Elicitation doesn't accept a whole bunch of otherwise valid POCOs for the same reason, so I don't see why would want to draw the line with nullables specifically. Consider also that the elicitation spec could evolve in the future allowing nullable properties, in which case we wouldn't be able to adapt to that without introducing breaking changes here.